File: web/src/engine/theory.ts
/**
 * Music Theory Module
 *
 * Ported from melody_architect/theory.py.
 * Implements pitch class utilities, Krumhansl-Schmuckler key-finding profiles,
 * Roman numeral parsing, chord resolution, and mode interval definitions.
 */

import type { Chord } from '../types/music';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Pitch class names using sharps. */
export const PITCH_CLASS_NAMES: readonly string[] = [
  'C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B',
];

/** Pitch class names using flats. */
export const PITCH_CLASS_NAMES_FLAT: readonly string[] = [
  'C', 'Db', 'D', 'Eb', 'E', 'F', 'Gb', 'G', 'Ab', 'A', 'Bb', 'B',
];

/** Krumhansl-Schmuckler major key profile weights. */
export const MAJOR_PROFILE: readonly number[] = [
  6.35, 2.23, 3.48, 2.33, 4.38, 4.09, 2.52, 5.19, 2.39, 3.66, 2.29, 2.88,
];

/** Krumhansl-Schmuckler minor key profile weights. */
export const MINOR_PROFILE: readonly number[] = [
  6.33, 2.68, 3.52, 5.38, 2.60, 3.53, 2.54, 4.75, 3.98, 2.69, 3.34, 3.17,
];

/** Semitone intervals for common modes/scales. */
export const MODE_INTERVALS: ReadonlyMap<string, readonly number[]> = new Map([
  ['major', [0, 2, 4, 5, 7, 9, 11]],
  ['minor', [0, 2, 3, 5, 7, 8, 10]],
  ['dorian', [0, 2, 3, 5, 7, 9, 10]],
  ['mixolydian', [0, 2, 4, 5, 7, 9, 10]],
]);

/** Mapping of upper-case Roman numeral strings to scale degrees (1-based). */
const ROMAN_TO_DEGREE: ReadonlyMap<string, number> = new Map([
  ['I', 1],
  ['II', 2],
  ['III', 3],
  ['IV', 4],
  ['V', 5],
  ['VI', 6],
  ['VII', 7],
]);

/** Interval sets for chord qualities (semitones above root). */
const CHORD_TONE_INTERVALS: ReadonlyMap<string, readonly number[]> = new Map([
  ['major', [0, 4, 7]],
  ['minor', [0, 3, 7]],
  ['dim', [0, 3, 6]],
  ['dominant7', [0, 4, 7, 10]],
  ['maj7', [0, 4, 7, 11]],
  ['min7', [0, 3, 7, 10]],
]);

/** Regex for parsing Roman numeral tokens such as "bVII7" or "ii7". */
const ROMAN_PATTERN = /^(?<accidental>[b#]?)(?<roman>[ivIV]+)(?<suffix>.*)$/;

// ---------------------------------------------------------------------------
// Parsed Roman Numeral
// ---------------------------------------------------------------------------

export interface ParsedRomanNumeral {
  degree: number;
  flat: boolean;
  sharp: boolean;
  suffix: string;
  loweredCase: boolean;
}

// ---------------------------------------------------------------------------
// Helper Functions
// ---------------------------------------------------------------------------

/**
 * Convert a pitch class (0-11) to its name, using sharps by default.
 */
export function pitchClassToName(pc: number, preferFlats = false): string {
  const names = preferFlats ? PITCH_CLASS_NAMES_FLAT : PITCH_CLASS_NAMES;
  return names[((pc % 12) + 12) % 12];
}

/**
 * Convert a note name (e.g. "C#", "Db") to its pitch class (0-11).
 * Throws if the name is not recognised.
 */
export function nameToPitchClass(name: string): number {
  const token = name.trim().replace('\u266D', 'b').replace('\u266F', '#');
  const sharpIdx = PITCH_CLASS_NAMES.indexOf(token);
  if (sharpIdx !== -1) return sharpIdx;
  const flatIdx = PITCH_CLASS_NAMES_FLAT.indexOf(token);
  if (flatIdx !== -1) return flatIdx;
  throw new Error(`Unsupported note name: ${name}`);
}

/**
 * Rotate an array by `shift` positions (used to align profiles with a tonic).
 */
export function rotate(values: readonly number[], shift: number): number[] {
  const size = values.length;
  return Array.from({ length: size }, (_, idx) => values[((idx - shift) % size + size) % size]);
}

/**
 * Cosine similarity between two numeric vectors of equal length.
 */
export function cosineSimilarity(a: readonly number[], b: readonly number[]): number {
  let dot = 0;
  let na = 0;
  let nb = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    na += a[i] * a[i];
    nb += b[i] * b[i];
  }
  na = Math.sqrt(na);
  nb = Math.sqrt(nb);
  if (na === 0 || nb === 0) return 0;
  return dot / (na * nb);
}

// ---------------------------------------------------------------------------
// Roman Numeral Parsing
// ---------------------------------------------------------------------------

/**
 * Parse a Roman numeral token (e.g. "bVII7", "ii7", "Imaj7") into its
 * constituent parts: scale degree, accidental, suffix, and case.
 */
export function parseRomanNumeral(token: string): ParsedRomanNumeral {
  const text = token.trim();
  const match = ROMAN_PATTERN.exec(text);
  if (!match || !match.groups) {
    throw new Error(`Invalid roman numeral token: ${token}`);
  }

  const accidentalPrefix = match.groups['accidental'];
  const roman = match.groups['roman'];
  const suffix = match.groups['suffix'] ?? '';

  const upperRoman = roman.toUpperCase();
  const degree = ROMAN_TO_DEGREE.get(upperRoman);
  if (degree === undefined) {
    throw new Error(`Unsupported roman numeral: ${token}`);
  }

  return {
    degree,
    flat: accidentalPrefix === 'b',
    sharp: accidentalPrefix === '#',
    suffix,
    loweredCase: roman === roman.toLowerCase(),
  };
}

// ---------------------------------------------------------------------------
// Chord Quality Inference
// ---------------------------------------------------------------------------

/**
 * Infer chord quality from a parsed Roman numeral token.
 */
function inferChordQuality(parsed: ParsedRomanNumeral): string {
  const suffix = parsed.suffix.toLowerCase();
  if (suffix.includes('dim') || suffix.includes('\u00B0')) return 'dim';
  if (suffix.includes('maj7')) return 'maj7';
  if (suffix.startsWith('m7')) return 'min7';
  if (suffix === '7') return parsed.loweredCase ? 'min7' : 'dominant7';
  if (parsed.loweredCase) return 'minor';
  return 'major';
}

// ---------------------------------------------------------------------------
// Chord Tone Retrieval
// ---------------------------------------------------------------------------

/**
 * Return the pitch classes of chord tones given a root pitch class and quality.
 */
export function getChordTones(rootPc: number, quality: string): number[] {
  const intervals = CHORD_TONE_INTERVALS.get(quality);
  if (!intervals) {
    throw new Error(`Unsupported chord quality: ${quality}`);
  }
  return intervals.map((interval) => (rootPc + interval) % 12);
}

// ---------------------------------------------------------------------------
// Scale Intervals
// ---------------------------------------------------------------------------

/**
 * Retrieve the semitone intervals for a given mode.
 */
export function scaleIntervals(mode: string): readonly number[] {
  const intervals = MODE_INTERVALS.get(mode);
  if (!intervals) {
    throw new Error(`Unsupported mode: ${mode}`);
  }
  return intervals;
}

// ---------------------------------------------------------------------------
// Chord Symbol Generation
// ---------------------------------------------------------------------------

/**
 * Build a human-readable chord symbol from a root name and quality.
 */
function chordSymbol(rootName: string, quality: string): string {
  switch (quality) {
    case 'major':
      return rootName;
    case 'minor':
      return `${rootName}m`;
    case 'dim':
      return `${rootName}dim`;
    case 'dominant7':
      return `${rootName}7`;
    case 'maj7':
      return `${rootName}maj7`;
    case 'min7':
      return `${rootName}m7`;
    default:
      throw new Error(`Unsupported chord quality: ${quality}`);
  }
}

// ---------------------------------------------------------------------------
// Resolve Roman Numeral to Chord
// ---------------------------------------------------------------------------

/**
 * Resolve a Roman numeral token (e.g. "IV", "bVII7") to a fully specified
 * Chord object, given a tonic pitch class and mode.
 */
export function resolveRomanToChord(token: string, tonicPc: number, mode: string): Chord {
  const parsed = parseRomanNumeral(token);
  const intervals = scaleIntervals(mode);
  const accidental = parsed.flat ? -1 : parsed.sharp ? 1 : 0;
  const rootPc = ((tonicPc + intervals[parsed.degree - 1] + accidental) % 12 + 12) % 12;
  const quality = inferChordQuality(parsed);
  const tones = getChordTones(rootPc, quality);
  const preferFlats = token.includes('b');
  const rootName = pitchClassToName(rootPc, preferFlats);
  return {
    rootPc,
    rootName,
    quality,
    symbol: chordSymbol(rootName, quality),
    roman: token,
    tones,
  };
}

File: melody_architect/theory.py
from __future__ import annotations

import math
import re
import sqlite3
from dataclasses import dataclass
from pathlib import Path

from .models import Chord

PC_NAMES_SHARP = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"]
PC_NAMES_FLAT = ["C", "Db", "D", "Eb", "E", "F", "Gb", "G", "Ab", "A", "Bb", "B"]

MAJOR_PROFILE = [6.35, 2.23, 3.48, 2.33, 4.38, 4.09, 2.52, 5.19, 2.39, 3.66, 2.29, 2.88]
MINOR_PROFILE = [6.33, 2.68, 3.52, 5.38, 2.60, 3.53, 2.54, 4.75, 3.98, 2.69, 3.34, 3.17]

MODE_INTERVALS = {
    "major": [0, 2, 4, 5, 7, 9, 11],
    "minor": [0, 2, 3, 5, 7, 8, 10],
    "dorian": [0, 2, 3, 5, 7, 9, 10],
    "mixolydian": [0, 2, 4, 5, 7, 9, 10],
    "phrygian": [0, 1, 3, 5, 7, 8, 10],
    "lydian": [0, 2, 4, 6, 7, 9, 11],
    "aeolian": [0, 2, 3, 5, 7, 8, 10],
    "locrian": [0, 1, 3, 5, 6, 8, 10],
    "harmonic_minor": [0, 2, 3, 5, 7, 8, 11],
    "melodic_minor": [0, 2, 3, 5, 7, 9, 11],
    "pentatonic_major": [0, 2, 4, 7, 9],
    "pentatonic_minor": [0, 3, 5, 7, 10],
    "blues": [0, 3, 5, 6, 7, 10],
    "whole_tone": [0, 2, 4, 6, 8, 10],
    "diminished_hw": [0, 1, 3, 4, 6, 7, 9, 10],
    "diminished_wh": [0, 2, 3, 5, 6, 8, 9, 11],
    "bebop_dominant": [0, 2, 4, 5, 7, 9, 10, 11],
    "phrygian_dominant": [0, 1, 4, 5, 7, 8, 10],
    "hungarian_minor": [0, 2, 3, 6, 7, 8, 11],
    "double_harmonic": [0, 1, 4, 5, 7, 8, 11],
}

ROMAN_TO_DEGREE = {
    "I": 1,
    "II": 2,
    "III": 3,
    "IV": 4,
    "V": 5,
    "VI": 6,
    "VII": 7,
}

CHORD_TONE_INTERVALS = {
    "major": [0, 4, 7],
    "minor": [0, 3, 7],
    "dim": [0, 3, 6],
    "dominant7": [0, 4, 7, 10],
    "maj7": [0, 4, 7, 11],
    "min7": [0, 3, 7, 10],
    "aug": [0, 4, 8],
    "sus2": [0, 2, 7],
    "sus4": [0, 5, 7],
    "dom7": [0, 4, 7, 10],
    "min_maj7": [0, 3, 7, 11],
    "dim7": [0, 3, 6, 9],
    "half_dim7": [0, 3, 6, 10],
    "aug7": [0, 4, 8, 10],
    "aug_maj7": [0, 4, 8, 11],
    "6": [0, 4, 7, 9],
    "min6": [0, 3, 7, 9],
    "9": [0, 4, 7, 10, 14],
    "maj9": [0, 4, 7, 11, 14],
    "min9": [0, 3, 7, 10, 14],
    "add9": [0, 4, 7, 14],
    "11": [0, 4, 7, 10, 14, 17],
    "min11": [0, 3, 7, 10, 14, 17],
    "13": [0, 4, 7, 10, 14, 17, 21],
    "power5": [0, 7],
    "7sharp5": [0, 4, 8, 10],
    "7flat5": [0, 4, 6, 10],
    "7sharp9": [0, 4, 7, 10, 15],
    "7flat9": [0, 4, 7, 10, 13],
    "7sharp11": [0, 4, 7, 10, 18],
}

ROMAN_PATTERN = re.compile(r"^(?P<accidental>[b#]?)(?P<roman>[ivIV]+)(?P<suffix>.*)$")


def _normalize_lookup_key(name: str) -> str:
    token = name.strip().lower().replace("♭", "b").replace("♯", "#")
    token = re.sub(r"[^a-z0-9#]+", "_", token)
    return token.strip("_")


def _parse_intervals(interval_text: str) -> list[int]:
    return [int(part.strip()) for part in interval_text.split(",") if part.strip()]


def _knowledge_db_path() -> Path:
    return Path(__file__).resolve().parent.parent / "knowledge" / "knowledge.db"


def load_scales_from_db() -> dict[str, list[int]]:
    db_path = _knowledge_db_path()
    if not db_path.exists():
        return {}
    with sqlite3.connect(db_path) as conn:
        rows = conn.execute(
            "SELECT name, intervals FROM scales_modes ORDER BY id"
        ).fetchall()
    return {
        _normalize_lookup_key(name): _parse_intervals(intervals)
        for name, intervals in rows
    }


def load_chords_from_db() -> dict[str, list[int]]:
    db_path = _knowledge_db_path()
    if not db_path.exists():
        return {}
    with sqlite3.connect(db_path) as conn:
        rows = conn.execute(
            "SELECT name, intervals FROM chord_types ORDER BY id"
        ).fetchall()
    return {
        _normalize_lookup_key(name): _parse_intervals(intervals)
        for name, intervals in rows
    }


def get_scale(name: str) -> list[int]:
    key = _normalize_lookup_key(name)
    if key in MODE_INTERVALS:
        return MODE_INTERVALS[key]
    scales_from_db = load_scales_from_db()
    if key in scales_from_db:
        return scales_from_db[key]
    raise ValueError(f"Unsupported scale: {name}")


def get_chord_type(name: str) -> list[int]:
    key = _normalize_lookup_key(name)
    if key in CHORD_TONE_INTERVALS:
        return CHORD_TONE_INTERVALS[key]
    chords_from_db = load_chords_from_db()
    if key in chords_from_db:
        return chords_from_db[key]
    raise ValueError(f"Unsupported chord type: {name}")


def pc_to_name(pc: int, prefer_flats: bool = False) -> str:
    names = PC_NAMES_FLAT if prefer_flats else PC_NAMES_SHARP
    return names[pc % 12]


def note_name_to_pc(name: str) -> int:
    token = name.strip().replace("♭", "b").replace("♯", "#")
    if token in PC_NAMES_SHARP:
        return PC_NAMES_SHARP.index(token)
    if token in PC_NAMES_FLAT:
        return PC_NAMES_FLAT.index(token)
    raise ValueError(f"Unsupported note name: {name}")


def rotate(values: list[float], shift: int) -> list[float]:
    size = len(values)
    return [values[(idx - shift) % size] for idx in range(size)]


def cosine_similarity(a: list[float], b: list[float]) -> float:
    dot = sum(x * y for x, y in zip(a, b))
    na = math.sqrt(sum(x * x for x in a))
    nb = math.sqrt(sum(x * x for x in b))
    if na == 0 or nb == 0:
        return 0.0
    return dot / (na * nb)


@dataclass(frozen=True)
class RomanToken:
    accidental: int
    degree: int
    roman: str
    suffix: str
    lowered_case: bool


def parse_roman(token: str) -> RomanToken:
    text = token.strip()
    match = ROMAN_PATTERN.match(text)
    if not match:
        raise ValueError(f"Invalid roman numeral token: {token}")

    accidental_prefix = match.group("accidental")
    accidental = (
        -1 if accidental_prefix == "b" else (1 if accidental_prefix == "#" else 0)
    )
    roman = match.group("roman")
    suffix = match.group("suffix") or ""

    upper_roman = roman.upper()
    if upper_roman not in ROMAN_TO_DEGREE:
        raise ValueError(f"Unsupported roman numeral: {token}")

    return RomanToken(
        accidental=accidental,
        degree=ROMAN_TO_DEGREE[upper_roman],
        roman=roman,
        suffix=suffix,
        lowered_case=roman.islower(),
    )


def infer_chord_quality(token: RomanToken) -> str:
    suffix = token.suffix.lower().strip()
    if "dim" in suffix or "°" in suffix:
        return "dim"
    if suffix == "sus2":
        return "sus2"
    if suffix == "sus4":
        return "sus4"
    if suffix == "aug":
        return "aug"
    if "maj7" in suffix:
        return "maj7"
    if suffix.startswith("m7"):
        return "min7"
    if suffix == "7":
        return "min7" if token.lowered_case else "dominant7"
    if token.lowered_case:
        return "minor"
    return "major"


def scale_intervals(mode: str) -> list[int]:
    try:
        return get_scale(mode)
    except ValueError as exc:
        raise ValueError(f"Unsupported mode: {mode}") from exc


def chord_symbol(root_name: str, quality: str) -> str:
    if quality == "major":
        return root_name
    if quality == "minor":
        return f"{root_name}m"
    if quality == "dim":
        return f"{root_name}dim"
    if quality == "aug":
        return f"{root_name}aug"
    if quality == "sus2":
        return f"{root_name}sus2"
    if quality == "sus4":
        return f"{root_name}sus4"
    if quality == "dominant7":
        return f"{root_name}7"
    if quality == "maj7":
        return f"{root_name}maj7"
    if quality == "min7":
        return f"{root_name}m7"
    raise ValueError(f"Unsupported chord quality: {quality}")


def resolve_roman_to_chord(token: str, tonic_pc: int, mode: str) -> Chord:
    parsed = parse_roman(token)
    intervals = scale_intervals(mode)
    root_pc = (tonic_pc + intervals[parsed.degree - 1] + parsed.accidental) % 12
    quality = infer_chord_quality(parsed)
    tones = tuple(
        (root_pc + interval) % 12 for interval in CHORD_TONE_INTERVALS[quality]
    )
    prefer_flats = "b" in token
    root_name = pc_to_name(root_pc, prefer_flats=prefer_flats)
    return Chord(
        root_pc=root_pc,
        root_name=root_name,
        quality=quality,
        symbol=chord_symbol(root_name, quality),
        roman=token,
        tones=tones,
    )
